File: app/src/main/java/be/digitalia/fosdem/utils/ICalendarWriter.java
package be.digitalia.fosdem.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.Closeable;
import java.io.IOException;

import okio.BufferedSink;

/**
 * Simple wrapper to write to iCalendar file format.
 */
public class ICalendarWriter implements Closeable {

	private static final String CRLF = "\r\n";

	private final BufferedSink sink;

	public ICalendarWriter(@NonNull BufferedSink sink) {
		this.sink = sink;
	}

	public void write(@NonNull String key, @Nullable String value) throws IOException {
		if (value != null) {
			sink.writeUtf8(key);
			sink.writeUtf8CodePoint(':');

			// Escape line break sequences
			final int length = value.length();
			int start = 0;
			int end = 0;
			while (end < length) {
				final char c = value.charAt(end);
				if (c == '\r' || c == '\n') {
					sink.writeUtf8(value, start, end);
					sink.writeUtf8(CRLF);
					sink.writeUtf8CodePoint(' ');
					do {
						end++;
					}
					while ((end < length) && (value.charAt(end) == '\r' || value.charAt(end) == '\n'));
					start = end;
				} else {
					end++;
				}
			}
			sink.writeUtf8(value, start, length);

			sink.writeUtf8(CRLF);
		}
	}

	@Override
	public void close() throws IOException {
		sink.close();
	}
}

File: app/src/main/java/be/digitalia/fosdem/utils/ICalendarWriter.kt
package be.digitalia.fosdem.utils

import okio.BufferedSink
import java.io.Closeable
import java.io.IOException

/**
 * Simple wrapper to write to iCalendar file format.
 */
class ICalendarWriter(private val sink: BufferedSink) : Closeable {

    @Throws(IOException::class)
    fun write(key: String, value: String?) {
        if (value != null) {
            with(sink) {
                writeUtf8(key)
                writeUtf8CodePoint(':'.toInt())
                // Escape line break sequences
                val length = value.length
                var start = 0
                var end = 0
                while (end < length) {
                    val c = value[end]
                    if (c == '\r' || c == '\n') {
                        writeUtf8(value, start, end - start)
                        writeUtf8(CRLF)
                        writeUtf8CodePoint(' '.toInt())
                        do {
                            end++
                        } while (end < length && (value[end] == '\r' || value[end] == '\n'))
                        start = end
                    } else {
                        end++
                    }
                }
                writeUtf8(value, start, end - start)
                writeUtf8(CRLF)
            }
        }
    }

    @Throws(IOException::class)
    override fun close() {
        sink.close()
    }

    companion object {
        private const val CRLF = "\r\n"
    }

}